libqtile/widget/imapwidget.py

```diff
@@ -75,6 +75,9 @@ class ImapWidget(base.ThreadPoolText):
     def __init__(self, **config):
         base.ThreadPoolText.__init__(self, "", **config)
         self.add_defaults(ImapWidget.defaults)
+        if self.user is None:
+            logger.critical("No username provided")
+            return
```

Comment:
```suggestion
            raise ConfigError("You must set the 'user' parameter for the IMAP widget.")
```
Returning here doesn't stop the widget from polling so it's better to raise an error (which will be caught when qtile starts and a warning will be displayed in the bar).

You'll need to add an import too:
```python
from libqtile.confreader import ConfigError
```

Reply:
Well, I don't mind doing that but it means that the test now crashes immediately:

```pytb
_____________________________________ ERROR collecting test/widgets/test_widget_init_configure.py _____________________________________
.tox/py310-x11/lib/python3.10/site-packages/_pytest/runner.py:341: in from_call
    result: Optional[TResult] = func()
.tox/py310-x11/lib/python3.10/site-packages/_pytest/runner.py:372: in <lambda>
    call = CallInfo.from_call(lambda: list(collector.collect()), "collect")
.tox/py310-x11/lib/python3.10/site-packages/_pytest/python.py:531: in collect
    self._inject_setup_module_fixture()
.tox/py310-x11/lib/python3.10/site-packages/_pytest/python.py:545: in _inject_setup_module_fixture
    self.obj, ("setUpModule", "setup_module")
.tox/py310-x11/lib/python3.10/site-packages/_pytest/python.py:310: in obj
    self._obj = obj = self._getobj()
.tox/py310-x11/lib/python3.10/site-packages/_pytest/python.py:528: in _getobj
    return self._importtestmodule()
.tox/py310-x11/lib/python3.10/site-packages/_pytest/python.py:617: in _importtestmodule
    mod = import_path(self.path, mode=importmode, root=self.config.rootpath)
.tox/py310-x11/lib/python3.10/site-packages/_pytest/pathlib.py:567: in import_path
    importlib.import_module(module_name)
/usr/lib/python3.10/importlib/__init__.py:126: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
<frozen importlib._bootstrap>:1050: in _gcd_import
    ???
<frozen importlib._bootstrap>:1027: in _find_and_load
    ???
<frozen importlib._bootstrap>:1006: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:688: in _load_unlocked
    ???
.tox/py310-x11/lib/python3.10/site-packages/_pytest/assertion/rewrite.py:186: in exec_module
    exec(co, module.__dict__)
test/widgets/test_widget_init_configure.py:117: in <module>
    [
test/widgets/test_widget_init_configure.py:120: in <listcomp>
    if param[0]().orientations in [ORIENTATION_BOTH, ORIENTATION_VERTICAL]
libqtile/widget/imapwidget.py:80: in __init__
    raise ConfigError("You must set the 'user' parameter for the IMAP widget.")
E   libqtile.confreader.ConfigError: You must set the 'user' parameter for the IMAP widget.
========================================================== warnings summary ===========================================================
test/widgets/test_tasklist.py:29
  /tmp/qtile/test/widgets/test_tasklist.py:29: PytestCollectionWarning: cannot collect test class 'TestTaskList' because it has a __init__ constructor (from: test/widgets/test_tasklist.py)
    class TestTaskList(TaskList):

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
======================================================= short test summary info =======================================================
ERROR test/widgets/test_widget_init_configure.py - libqtile.confreader.ConfigError: You must set the 'user' parameter for the IMAP widget.
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
=============================================== 1 skipped, 1 warning, 1 error in 3.73s ================================================

```

Reply:
Hmm, I suppose I can add it to `no_test` list.

Reply:
Updated.